contributing/building-a-bento-video-player.md

```diff
@@ -0,0 +1,627 @@
+    [videoId]
+  );
+  // ...
+  return (
+    <VideoIframe
+      {...rest}
+      src={src}
+      ...
+    />
+  );
+}
+```
+
+#### `origin`
+
+By default, messages from the iframe are only verified by comparing their `contentWindow`. You should define a regular expression that verifies their origin:
+
+```js
+return (
+  <VideoIframe
+    {...rest}
+    origin={new RegExp('^https://((player|www)\.)?example\.com/?$')}
+    ...
+  />
+);
+```
+
+#### Playback methods with `makeMethodMessage`
+
+We need the ability to tell the iframe to execute certain actions, for example `play`, `mute` or `hideControls`. When using an iframe to load a player, this is done by sending a `postMessage` downstream.
+
+`makeMethodMessage` takes an action to execute as a `string`, and returns another `string` corresponding to a message to send.
+
+```ts
+type MakeMethodMessageFunction = (method: string) => string;
+```
+
+We implement this function with `useCallback()` so that it's created only when required as specified by hook dependencies. It's recommended that you also create a higher-level `makeMessage` function that creates and serializes messages as your iframe's interface needs it. In this case, we `JSON.stringify` a `videoId` and a `method`:
+
+```js
+function makeMessage(videoId, method) {
+  return JSON.stringify({
+    'videoId': videoId,
+    'method': method,
+  });
+}
+
+function FantasticPlayerWithRef(
+  {videoId, ...rest},
+  ref
+) {
+  // ...
+  const makeMethodMessage = useCallback(
+    (method) => makeMessage(videoId, method),
+    [videoId]
+  );
+
+  // ...
+  return (
+    <VideoIframe
+      {...rest}
+      makeMethodMessage={makeMethodMessage}
+      ...
+    />
+  );
+}
+```
+
+#### Handling events with `onMessage`
+
+Upstream events originated by the iframe are received as messages. You should define a function that interprets these messages and dispatches [`HTMLMediaElement` events](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement#events).
+
+```ts
+type OnMessageFunction = (event: MessageEvent) => void;
+```
+
+> ⚠️ **This is an incomplete [`MessageEvent`](https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent).** It's a copy that contains `currentTarget` (the originating `<iframe>`), `target` (same as `currentTarget`) and `data` (the data sent by the iframe with `postMessage`). If you need to copy other properties, add them to the appropriate place in [`video-iframe.js`](../extensions/amp-video/1.0/video-iframe.js) and add `@alanorozco` as a reviewer on your pull request.
+
+Here we implement the `canplay`, `play` and `pause` events for an iframe that posts them as the message `{"event": "play"}`.
+
+```js
+function onMessage(event) {
+  const {data, currentTarget} = event;
+  switch (data?.event) {
+    case 'canplay':
+    case 'play':
+    case 'pause':
+      dispatchCustomEvent(currentTarget, data.event, null, {
+        bubbles: true,
+        cancelable: false,
```

Comment:
Do bubbles/cancelable need to be wrapped in quotes? I have a hard time keeping track

Reply:
Per type check, no, since we only want an `Object`.

This doesn't cause any issues since we don't do advanced compilation, and we cannot do it without singlepass. Does it even matter anymore?